File: tutorial/lru.py
#!/usr/bin/env python3

class DLinkedNode(): 
    def __init__(self):
        self.key = 0
        self.value = 0
        self.prev = None
        self.next = None

class LRUCache():
    def __init__(self, capacity):
        self.cache = {}
        self.size = 0
        self.capacity = capacity
        self.head, self.tail = DLinkedNode(), DLinkedNode()
        self.head.next = self.tail
        self.tail.prev = self.head

    def _add_node(self, node):
        node.prev = self.head
        node.next = self.head.next
        self.head.next.prev = node
        self.head.next = node

    def _remove_node(self, node):
        prev = node.prev
        new = node.next
        prev.next = new
        new.prev = prev

    def _move_to_head(self, node):
        self._remove_node(node)
        self._add_node(node)

    def _pop_tail(self):
        node = self.tail.prev
        self._remove_node(node)
        return node

    def get(self, key):
        node = self.cache.get(key, None)
        if not node:
            return None
        self._move_to_head(node)
        return node.value

    def put(self, key, value):
        node = self.cache.get(key)

        if not node:
            new = DLinkedNode()
            new.key = key
            new.value = value
            self.cache[key] = new
            self._add_node(new)
            self.size += 1
            if self.size > self.capacity:
                tail = self._pop_tail()
                del self.cache[tail.key]
                self.size -= 1
        else:
            node.value = value
            self._move_to_head(node)

# Input
# ["LRUCache", "put", "put", "get", "put", "get", "put", "get", "get", "get"]
# [[2], [1, 1], [2, 2], [1], [3, 3], [2], [4, 4], [1], [3], [4]]
# Output
# [null, null, null, 1, null, -1, null, -1, 3, 4]

lru = LRUCache(2)
lru.put(1, 1)
lru.put(2, 2)
print(lru.get(1))
lru.put(3, 3)
print(lru.get(2))
lru.put(4, 4)
print(lru.get(1))
print(lru.get(3))
print(lru.get(4))

File: tutorial/lru.swift
#!/usr/bin/swift

class Node<Key: Hashable, Value> {
    var value: Value?
    var key: Key?
    var next: Node?
    var prev: Node?
}

struct LRUCache<Key: Hashable, Value> {
    typealias NodeType = Node<Key, Value>
    var cache = [Key: NodeType]()
    var size = 0
    let capacity: Int
    let head = NodeType()
    let tail = NodeType()
    init(capacity: Int) {
        self.capacity = capacity
        head.next = tail
        tail.prev = head
    }

    private func add(_ node: NodeType) {
        node.prev = head
        node.next = head.next
        head.next?.prev = node
        head.next = node
    }

    private func remove(_ node: NodeType) {
        guard let prev = node.prev, let next = node.next else {
            return
        }
        prev.next = next
        next.prev = prev
    }

    private func moveToHead(_ node: NodeType) {
        remove(node)
        add(node)
    }

    private func pop() -> NodeType? {
        guard let node = tail.prev, node !== head else {
            return nil
        }
        remove(node)
        return node
    }

    func value(for key: Key) -> Value? {
        guard let node = cache[key] else {
            return nil
        }
        moveToHead(node)
        return node.value
    }
}
